src/babel/transformation/transformers/es6/parameters.rest.js

```diff
@@ -17,6 +70,35 @@ exports.Function = function (node, parent, scope) {
   // otherwise `arguments` will be remapped in arrow functions
   argsId._ignoreAliasFunctions = true;
 
+  // support patterns
+  if (t.isPattern(rest)) {
+    var pattern = rest;
+    rest = scope.generateUidIdentifier("ref");
+    var declar = t.variableDeclaration("var", pattern.elements.map(function (elem, index) {
+      var accessExpr = t.memberExpression(rest, t.literal(index));
+      return t.variableDeclarator(elem, accessExpr);
+    }));
+    node.body.body.unshift(declar);
+  }
+
+  // check if rest is used only in member expressions
+  var restOuterDeclar = scope.getBindingIdentifier(rest.name);
+  var state = {
+    name: rest.name,
+    outerDeclar: restOuterDeclar,
+    isOptimizable: true,
+    candidates: [],
+    strictMode: file.transformers.useStrict.canRun()
+  };
+  scope.traverse(node, memberExpressionVisitor, state);
+
+  if (state.isOptimizable) {
+    for (let candidate of state.candidates) {
```

Comment:
I hoped, that cost of for-of is acceptable in comparison with scope traversal (that is right above).
Fixed. I'll push this after I investigate what is wrong with esnext tests for rest params.